amm/contracts/router_v2/stable_pool.rs

```diff
@@ -31,6 +39,13 @@ impl StablePool {
             },
             Err(_) => return None,
         };
+        // set spending allowance of each token for the pool to `u128::MAX`
+        // required for adding liquidity
+        for &token in &tokens {
+            if psp22_approve(token, pool_id, u128::MAX).is_err() {
+                return None;
+            };
+        }
```

Comment:
It is necessary to make these approvals inside the contract (`Router` approves `StableSwap` to spend its tokens). The `StableSwap::add_liquidity(...)` uses `PSP22::transferFrom(...)` to transfer tokens from the caller (in this case the caller is the `Router` contract). So basically the flow is:

1. User approves `Router` to spend their tokens 
2. `Router` transfers tokens from the user to itself
3. `Router` approves `StablePool` to spend its tokens
4. `StablePool` transfers tokens from `Router` to itself

To avoid doing Step 3. with every `add_stable_pool_liqudity(...)` call, it is done only once, when the `StablePool` is cached.

It is not necessary for the `Pair` because `Pair::mint(...)` assumes that the tokens are already transferred (amounts = balances - reserves). So the flow there looks like this:

1. User approves `Router` to spend their tokens 
2. (Optional) If "native" then `Router` transfers tokens to itself
3. `Router` transfers tokens from the user (or itself if Step 2.) to the `Pair`
4. `Pair` calls each token contract the get balance and computes the transferred amounts

Reply:
Ah makes sense, it's good then.